tailnet/multiagent.go

```diff
@@ -0,0 +1,120 @@
+package tailnet
+
+import (
+	"context"
+	"sync"
+
+	"github.com/google/uuid"
+
+	"cdr.dev/slog"
+)
+
+type MultiAgentConn interface {
+	UpdateSelf(node *Node) error
+	SubscribeAgent(agentID uuid.UUID, node *Node) error
+	UnsubscribeAgent(agentID uuid.UUID)
+	NextUpdate(ctx context.Context) []AgentNode
+	AgentIsLegacy(agentID uuid.UUID) bool
+	Close() error
+}
+
+type MultiAgent struct {
+	mu sync.RWMutex
+
+	ID     uuid.UUID
+	Logger slog.Logger
+
+	AgentIsLegacyFunc func(agentID uuid.UUID) bool
+	OnSubscribe       func(id uuid.UUID, agent uuid.UUID, node *Node) error
+	OnNodeUpdate      func(id uuid.UUID, agents []uuid.UUID, node *Node) error
+	OnClose           func(id uuid.UUID)
+
+	updates          chan AgentNode
+	subscribedAgents map[uuid.UUID]struct{}
+}
+
+type AgentNode struct {
+	AgentID uuid.UUID
+	*Node
+}
+
+func (m *MultiAgent) Init() *MultiAgent {
+	m.updates = make(chan AgentNode, 128)
+	m.subscribedAgents = map[uuid.UUID]struct{}{}
+	return m
+}
+
+func (m *MultiAgent) AgentIsLegacy(agentID uuid.UUID) bool {
+	return m.AgentIsLegacyFunc(agentID)
+}
+
+func (m *MultiAgent) OnAgentUpdate(id uuid.UUID, node *Node) {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+
+	if _, ok := m.subscribedAgents[id]; !ok {
+		return
+	}
+
+	select {
+	case m.updates <- AgentNode{AgentID: id, Node: node}:
```

Comment:
it looks like this only gets called when an agent publishes a new node, but the MultiAgentConn wants to get the latest node bindings for agents as soon as they subscribe---not wait until the agent publishes an update, which could be never, if its endpoints and DERP home haven't changed.